### DISCUSSION TITLE
Texture drop logic

### DISCUSSION THREAD

**Discussion by raphamorim** [ayosec/alacritty#12] [Q&A] [answered]
Hey @ayosec thanks for the awesome work you have done in Alacritty and for making it available for any person that likes alacritty.

I have an question regarding your implementation, I ported most of your changes to Rio terminal https://github.com/raphamorim/rio but I can see that you added a logic for texture drop then graphic removal, I would like to confirm about if there's anything on term code that applies drop for it or is basically if it's removed from the row.

Asking because I found this issue in Rio implementation https://github.com/raphamorim/rio/issues/709 but I cannot reproduce with your terminal fork.

Thanks in advance 🙏

**Comment by ayosec** [comment, marked as answer]
> I would like to confirm about if there's anything on term code that applies drop for it or is basically if it's removed from the row.

I'm not sure if I understand the question.

In this patch, there are two types related to graphics in the terminal:

* [`TextureRef`](https://github.com/ayosec/alacritty/blob/ecf278007a5fe024fe624c7dea4664239d81f44b/alacritty_terminal/src/graphics/mod.rs#L40-L55) contains a reference to the texture stored in the GPU.
* [`GraphicCell`](https://github.com/ayosec/alacritty/blob/ecf278007a5fe024fe624c7dea4664239d81f44b/alacritty_terminal/src/graphics/mod.rs#L79-L91) is included in every cell in the terminal. Each instance shares a reference to the `TextureRef` of the texture.

`GraphicCell` is dropped when anything is written to the cell. In [its implementation](https://github.com/ayosec/alacritty/blob/ecf278007a5fe024fe624c7dea4664239d81f44b/alacritty_terminal/src/graphics/mod.rs#L102-L114), it sends a `ClearSubregion` operation to fill with transparent pixels the replaced cell.

`TextureRef` is dropped when all `GraphicCell` are dropped. The [implementation](https://github.com/ayosec/alacritty/blob/ecf278007a5fe024fe624c7dea4664239d81f44b/alacritty_terminal/src/graphics/mod.rs#L66-L72) sends a `Remove` operation to release the GPU memory associated to the texture.

**Comment by raphamorim** [comment]
I see! Thank you for replying. Sorry the question was a bit confusing, I just wanted to understand if I was missing any drop call that wasn't in PR implementation you did originally
